src/algebra/direct_sum_group.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2019 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+
+Direct sum of abelian groups, indexed by a discrete type.
+-/
+import algebra.direct_sum
+
+open_locale big_operators
+
+universes u v w u₁
+
+namespace direct_sum
+
+variables {ι : Type v} [decidable_eq ι]
+variables {β : ι → Type w} [Π i, add_comm_group (β i)]
+
+instance : add_comm_group (direct_sum ι β) :=
+dfinsupp.add_comm_group
+
+instance mk.is_add_group_hom (s : finset ι) : is_add_group_hom (mk β s) :=
+{ map_add := λ _ _, dfinsupp.mk_add }
+
+@[simp] lemma mk_neg (s : finset ι) (x) : mk β s (-x) = -mk β s x :=
+is_add_group_hom.map_neg _ x
+
+@[simp] lemma mk_sub (s : finset ι) (x y) : mk β s (x - y) = mk β s x - mk β s y :=
+is_add_group_hom.map_sub _ x y
+
+instance of.is_add_group_hom (i : ι) : is_add_group_hom (of β i) :=
+{ map_add := λ _ _, dfinsupp.single_add }
+
+@[simp] lemma of_neg (i : ι) (x) : of β i (-x) = -of β i x :=
+is_add_group_hom.map_neg _ x
+
+@[simp] lemma of_sub (i : ι) (x y) : of β i (x - y) = of β i x - of β i y :=
+is_add_group_hom.map_sub _ x y
+
+variables {γ : Type u₁} [add_comm_group γ]
+variables (φ : Π i, β i → γ) [Π i, is_add_group_hom (φ i)]
+
+variables (φ)
+def to_group (f : direct_sum ι β) : γ := @to_add_monoid _ _ _ _ _ _ φ (λ i, by apply_instance) f
```

Comment:
This `(λ i, by apply_instance)` is annoying. Confusingly, `(λ i, infer_instance)` does not work.